File: leetcode/Priority_Queue/0373_Find_K_Pairs_with_Smallest_Sums.cpp
#include <utility>
#include <algorithm>

class mycomparison{
private:
	vector<int> nums1;
	vector<int> nums2;
public:
	mycomparison(vector<int> &n1, vector<int> &n2) {
		this -> nums1 = n1;
		this -> nums2 = n2;
	}
	bool operator() (const pair<int, int>& p1, const pair<int, int>& p2) {
		return (nums1[p1.first] + nums2[p1.second] > nums1[p2.first] + nums2[p2.second]);
	}
};

class Solution {
public:
    vector<pair<int, int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        vector<pair<int, int>> result;
        if (nums1.size() == 0 || nums2.size() == 0 || k == 0)
        	return result;

        priority_queue<pair<int, int>, vector<pair<int, int>>, mycomparison> pq(mycomparison(nums1, nums2));
        int n = min(k, (int)nums1.size() * (int)nums2.size());

        for (int i = 0; i < min(k, (int)nums1.size()); i++)
        	pq.emplace(make_pair(i, 0));

        for (int i = 0; i < n; i++) {
        	pair<int, int> next = pq.top();
        	pq.pop();
        	result.emplace_back(make_pair(nums1[next.first], nums2[next.second]));
        	if (next.second < nums2.size() - 1)
        		pq.emplace(make_pair(next.first, next.second + 1));
        }

        return result;
    }
};

File: leetcode/Priority_Queue/0373_Find_K_Pairs_with_Smallest_Sums.py
class Solution(object):
    def kSmallestPairs(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: List[List[int]]
        """
        N, M = len(nums1), len(nums2)
        # for each num in nums1, record the current index of corresponding num in nums2
        # and maintain by non-decreasing sum val
        pq = [(nums1[i] + nums2[0], i, 0) for i in range(N)] # list of tuple (sum val, nums1 index, nums2 index)
        result = []
        for _ in range(k):
            _, i, j = pq.pop(0)
            result.append([nums1[i], nums2[j]])
            if j == M - 1:
                continue
            else:
                new_tuple = (nums1[i] + nums2[j+1], i, j+1)
                # use binary search to insert new tuple back to pq
                left, right = 0, len(pq) - 1
                while left <= right:
                    mid = (left + right) // 2
                    if pq[mid][0] <= new_tuple[0]:
                        left = mid + 1
                    else:
                        right = mid - 1
                # final L is min element index which > target, i.e. inserting point
                pq.insert(left, new_tuple)

        return result
